homeassistant/components/homekit/switches.py

```diff
@@ -0,0 +1,63 @@
+"""Class to hold all switch accessories."""
+import logging
+
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory
+from .const import (SERV_SWITCH, CHAR_ON)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@TYPES.register('Switch')
+class Switch(HomeAccessory):
+    """Generate a Switch accessory."""
+
+    def __init__(self, hass, entity_id, display_name):
+        """Initialize a Switch accessory object to represent a remote."""
+        super().__init__(display_name, entity_id, 'SWITCH')
+
+        self._hass = hass
+        self._entity_id = entity_id
+        self._domain = hass.states.get(self._entity_id).domain
+
+        self.current_on = None
+        self.homekit_target_on = None
+
+        self.service_switch = self.get_service(SERV_SWITCH)
+        self.char_on = self.service_switch.get_characteristic(CHAR_ON)
+        self.char_on.setter_callback = self.set_state
+
+    def run(self):
+        """Method called be object after driver is started."""
+        state = self._hass.states.get(self._entity_id)
+        self.update_state(new_state=state)
+
+        async_track_state_change(self._hass, self._entity_id,
+                                 self.update_state)
+
+    def set_state(self, value):
+        """Move switch state to value if call came from homekit."""
+        if value != self.current_on:
+            _LOGGER.debug("%s: Set switch state to %s",
+                          self._entity_id, value)
+            self.homekit_target_on = value
+            service = 'turn_on' if value else 'turn_off'
+            self._hass.services.call(self._domain, service,
+                                     {'entity_id': self._entity_id})
+
+    def update_state(self, entity_id=None, old_state=None, new_state=None):
+        """Update switch state after state changed."""
+        if new_state is None:
+            return
+
```

Comment:
Probably I'm missing something, but how can I know if the call came from Home Assistant or not? Furthermore, as far as I can see, update_state is only called from Switch.run and through the handler of the track state change, so it can only be called through Home Assistant, right?

Reply:
No quite. A `state_changed` event is also called if HomeKit wants to set the state value for Home Assistant. There a two cases in which `update_state` is called:

* `HomeKit value change` -> `HA state_change` -> `update_state`
* `HA value change` -> `HA state_change` -> `update_state`

During my first review I suggested to use `should_callback=False`. This makes sure that `set_state` is only called from HomeKit. Sadly there is no such option for the Home Assistant side. The best solution I have come up with so far is to set a flag var in `set_state` and check/remove it during `update_state`.

Reply:
Ok, I see. I guess the only "problem" is performing the set_value on the characteristic? I now only do it when the call did not come from HomeKit (i.e. when homekit_target_on is None). The remainder of the update_state is needed in my opinion to have current_on set correctly